frontend/src/scenes/session-recordings/player/sessionRecordingPlayerLogic.ts

```diff
@@ -179,86 +183,90 @@ export const sessionRecordingPlayerLogic = kea<sessionRecordingPlayerLogicType>(
-            const currentEvents = values.replayer?.service.state.context.events ?? []
-            const eventsToAdd = values.snapshots.slice(currentEvents.length) ?? []
+            const currentEvents = values.player?.replayer?.service.state.context.events ?? []
+            const eventsToAdd = []
+            if (values.currentSegment?.windowId) {
+                eventsToAdd.push(
+                    ...(values.sessionPlayerData.snapshotsByWindowId[values.currentSegment?.windowId] ?? []).slice(
+                        currentEvents.length
+                    )
+                )
+            }
 
             // If replayer isn't initialized, it will be initialized with the already loaded snapshots
-            if (!!values.replayer) {
+            if (!!values.player?.replayer) {
                 eventsToAdd.forEach((event: eventWithTime) => {
-                    values.replayer?.addEvent(event)
+                    values.player?.replayer?.addEvent(event)
                 })
+            } else {
+                actions.tryInitReplayer()
             }
 
-            // replayer should be updated with new events.
-            const finalEvents = [...currentEvents, ...eventsToAdd]
-            if (finalEvents.length < 1) {
-                return
-            }
-
-            // Update last buffered point
-            const lastEvent = finalEvents[finalEvents.length - 1]
-            actions.setLastBufferedTime(lastEvent.timestamp)
-
             // If buffering has completed, resume last playing state
             if (
+                values.currentPlayerPosition &&
+                values.sessionPlayerData.bufferedTo &&
                 values.currentPlayerState === SessionPlayerState.BUFFER &&
-                values.time.current + BUFFER_TIME_BUFFER_MS < lastEvent.timestamp
+                comparePlayerPositions(
+                    values.currentPlayerPosition,
+                    values.sessionPlayerData.bufferedTo,
+                    values.sessionPlayerData.metadata.segments
+                ) < 0
             ) {
                 actions.clearLoadingState()
                 actions.setPlay()
             }
 
-            // Set meta once whole session recording loads. This overrides the meta set when metadata
-            // was fetched separately.
-            if (!values.sessionPlayerDataLoading && !!values.replayer) {
-                const meta = values.replayer.getMetaData()
-                // Sometimes replayer doesn't update with events we recently added.
-                const endTime = Math.max(
-                    meta.endTime,
-                    finalEvents.length ? finalEvents[finalEvents.length - 1]?.timestamp : 0
-                )
-                const finalMeta = {
-                    ...meta,
-                    endTime,
-                    totalTime: endTime - meta.startTime,
-                }
-                actions.setMeta(finalMeta)
-            }
-
             breakpoint()
```

Comment:
I believe this second half of this buffer check should also be triggered on successful load of metadata (maybe via sharedListener), in the *rare* case that meta loads after a few snapshot chunks do.

Reply:
Good catch!